File: frontend/src/payroll/businessEntity.js
/** Business entity model — mirrors backend/business_entity.py */

export const BUSINESS_ENTITY = {
  WASHPRO: "washpro",
  WASHMATE: "washmate",
  VEEWASH: "veewash",
  RINSE_EXCLUSIVE: "rinse_exclusive",
  SHARED: "shared",
  NONE: "none",
  COMBINED: "combined",
};

export const ENTITY_LABELS = {
  [BUSINESS_ENTITY.WASHPRO]: "WashPro",
  [BUSINESS_ENTITY.WASHMATE]: "WashMate",
  [BUSINESS_ENTITY.VEEWASH]: "VeeWash",
  [BUSINESS_ENTITY.RINSE_EXCLUSIVE]: "Rinse Exclusive",
  [BUSINESS_ENTITY.SHARED]: "Shared",
  [BUSINESS_ENTITY.NONE]: "None",
  [BUSINESS_ENTITY.COMBINED]: "Combined (Admin)",
};

export const WORKER_ENTITY_OPTIONS = [
  { value: BUSINESS_ENTITY.WASHPRO, label: ENTITY_LABELS[BUSINESS_ENTITY.WASHPRO] },
  { value: BUSINESS_ENTITY.WASHMATE, label: ENTITY_LABELS[BUSINESS_ENTITY.WASHMATE] },
  { value: BUSINESS_ENTITY.VEEWASH, label: ENTITY_LABELS[BUSINESS_ENTITY.VEEWASH] },
  { value: BUSINESS_ENTITY.RINSE_EXCLUSIVE, label: ENTITY_LABELS[BUSINESS_ENTITY.RINSE_EXCLUSIVE] },
  { value: BUSINESS_ENTITY.SHARED, label: ENTITY_LABELS[BUSINESS_ENTITY.SHARED] },
  { value: BUSINESS_ENTITY.NONE, label: ENTITY_LABELS[BUSINESS_ENTITY.NONE] },
];

export function normalizeOrgSlug(raw) {
  return String(raw || "washpro").trim().toLowerCase() || "washpro";
}

export function defaultEntityForOrg(orgSlug) {
  const slug = normalizeOrgSlug(orgSlug);
  if (slug === "washmate") return BUSINESS_ENTITY.WASHMATE;
  if (slug === "veewash") return BUSINESS_ENTITY.VEEWASH;
  return BUSINESS_ENTITY.WASHPRO;
}

export function normalizeWorkerEntity(raw, organizationSlug = null) {
  const aff = String(raw || "").trim().toLowerCase();
  if (aff === "both") return BUSINESS_ENTITY.SHARED;
  if (Object.values(BUSINESS_ENTITY).includes(aff)) return aff;
  return null;
}

export function normalizeShiftEntity(raw, organizationSlug = null) {
  const aff = String(raw || "").trim().toLowerCase();
  if (aff === "veewash") {
    return normalizeOrgSlug(organizationSlug) === "veewash"
      ? BUSINESS_ENTITY.VEEWASH
      : BUSINESS_ENTITY.WASHPRO;
  }
  if (
    aff === BUSINESS_ENTITY.WASHPRO ||
    aff === BUSINESS_ENTITY.WASHMATE ||
    aff === BUSINESS_ENTITY.VEEWASH ||
    aff === BUSINESS_ENTITY.RINSE_EXCLUSIVE
  ) {
    return aff;
  }
  return null;
}

export function entityLabel(entity) {
  return ENTITY_LABELS[String(entity || "").toLowerCase()] || entity || "—";
}

/** Worker entity options allowed for this tenant login. */
export function workerEntityOptionsForOrganization(orgSlug) {
  const slug = normalizeOrgSlug(orgSlug);
  if (slug === "washmate") {
    return WORKER_ENTITY_OPTIONS.filter((opt) =>
      [BUSINESS_ENTITY.WASHMATE, BUSINESS_ENTITY.SHARED, BUSINESS_ENTITY.NONE].includes(opt.value),
    );
  }
  if (slug === "veewash") {
    return WORKER_ENTITY_OPTIONS.filter((opt) =>
      [BUSINESS_ENTITY.VEEWASH, BUSINESS_ENTITY.SHARED, BUSINESS_ENTITY.NONE].includes(opt.value),
    );
  }
  return WORKER_ENTITY_OPTIONS.filter((opt) =>
    [
      BUSINESS_ENTITY.WASHPRO,
      BUSINESS_ENTITY.RINSE_EXCLUSIVE,
      BUSINESS_ENTITY.SHARED,
      BUSINESS_ENTITY.NONE,
    ].includes(opt.value),
  );
}

export function entitySummaryTitle(orgSlug) {
  const slug = normalizeOrgSlug(orgSlug);
  if (slug === "washmate") return "WashMate · Shared · None";
  if (slug === "veewash") return "VeeWash · Shared · None";
  return "WashPro · Rinse Exclusive · Shared · None";
}

export function resolveStoredWorkerEntity(row, organizationSlug = null) {
  const raw = row?.business_entity || row?.employer_affiliation;
  const normalized = normalizeWorkerEntity(raw, organizationSlug);
  if (normalized) return normalized;
  return null;
}

export function entitiesForOrganization(orgSlug, { isPrivileged = false } = {}) {
  const slug = normalizeOrgSlug(orgSlug);
  let tabs;
  if (slug === "washmate") tabs = [BUSINESS_ENTITY.WASHMATE];
  else if (slug === "veewash") tabs = [BUSINESS_ENTITY.VEEWASH];
  else tabs = [BUSINESS_ENTITY.WASHPRO, BUSINESS_ENTITY.RINSE_EXCLUSIVE];
  if (isPrivileged) tabs = [...tabs, BUSINESS_ENTITY.COMBINED];
  return tabs;
}

export function workerMatchesEntityTab(workerEntity, tab, organizationSlug = null) {
  const tabKey = String(tab || "").toLowerCase();
  const entity = normalizeWorkerEntity(workerEntity, organizationSlug) || BUSINESS_ENTITY.NONE;
  if (tabKey === BUSINESS_ENTITY.COMBINED) return entity !== BUSINESS_ENTITY.NONE;
  if (entity === BUSINESS_ENTITY.NONE) return false;
  if (entity === BUSINESS_ENTITY.SHARED) {
    return entitiesForOrganization(organizationSlug).includes(tabKey);
  }
  return entity === tabKey;
}

export function shiftMatchesEntityTab(shiftEntity, tab, organizationSlug = null) {
  const tabKey = String(tab || "").toLowerCase();
  const entity = normalizeShiftEntity(shiftEntity, organizationSlug);
  if (tabKey === BUSINESS_ENTITY.COMBINED) return Boolean(entity);
  return entity === tabKey;
}

File: backend/business_entity.py
"""Business entity isolation — WashPro, WashMate, and VeeWash share a platform, not records."""

from __future__ import annotations

from typing import Any, Mapping, Sequence

ENTITY_WASHPRO = "washpro"
ENTITY_WASHMATE = "washmate"
ENTITY_VEEWASH = "veewash"
ENTITY_RINSE_EXCLUSIVE = "rinse_exclusive"
ENTITY_SHARED = "shared"
ENTITY_NONE = "none"
ENTITY_COMBINED = "combined"

ENTITY_LABELS: dict[str, str] = {
    ENTITY_WASHPRO: "WashPro",
    ENTITY_WASHMATE: "WashMate",
    ENTITY_VEEWASH: "VeeWash",
    ENTITY_RINSE_EXCLUSIVE: "Rinse Exclusive",
    ENTITY_SHARED: "Shared",
    ENTITY_NONE: "None",
    ENTITY_COMBINED: "Combined (Admin)",
}

WORKER_ENTITIES = frozenset(
    {
        ENTITY_WASHPRO,
        ENTITY_WASHMATE,
        ENTITY_VEEWASH,
        ENTITY_RINSE_EXCLUSIVE,
        ENTITY_SHARED,
        ENTITY_NONE,
    }
)
SHIFT_ENTITIES = frozenset(
    {
        ENTITY_WASHPRO,
        ENTITY_WASHMATE,
        ENTITY_VEEWASH,
        ENTITY_RINSE_EXCLUSIVE,
    }
)

ORG_SLUG_DEFAULT_ENTITY: dict[str, str] = {
    "washpro": ENTITY_WASHPRO,
    "washmate": ENTITY_WASHMATE,
    "veewash": ENTITY_VEEWASH,
}

PRIVILEGED_ROLES = frozenset({"ADMIN", "OPS", "SUPER_ADMIN", "PLATFORM_ADMIN"})


def normalize_org_slug(raw: Any) -> str:
    return str(raw or "washpro").strip().lower() or "washpro"


def default_entity_for_org(org_slug: Any) -> str:
    return ORG_SLUG_DEFAULT_ENTITY.get(normalize_org_slug(org_slug), ENTITY_WASHPRO)


def normalize_worker_entity(raw: Any, *, organization_slug: str | None = None) -> str | None:
    aff = str(raw or "").strip().lower()
    if aff == "both":
        return ENTITY_SHARED
    if aff in WORKER_ENTITIES:
        return aff
    return None


def normalize_shift_entity(raw: Any, *, organization_slug: str | None = None) -> str | None:
    aff = str(raw or "").strip().lower()
    if aff == "veewash":
        slug = normalize_org_slug(organization_slug)
        return ENTITY_VEEWASH if slug == "veewash" else ENTITY_WASHPRO
    if aff in SHIFT_ENTITIES:
        return aff
    if aff == ENTITY_RINSE_EXCLUSIVE:
        return ENTITY_RINSE_EXCLUSIVE
    return None


def entity_label(entity: str | None) -> str:
    key = str(entity or "").strip().lower()
    return ENTITY_LABELS.get(key, key or "—")


def entities_for_organization(org_slug: Any, *, is_privileged: bool = False) -> list[str]:
    """Entity tabs visible for a tenant login."""
    slug = normalize_org_slug(org_slug)
    if slug == "washmate":
        tabs = [ENTITY_WASHMATE]
    elif slug == "veewash":
        tabs = [ENTITY_VEEWASH]
    else:
        tabs = [ENTITY_WASHPRO, ENTITY_RINSE_EXCLUSIVE]
    if is_privileged:
        tabs = [*tabs, ENTITY_COMBINED]
    return tabs


def worker_matches_entity_tab(
    worker_entity: str | None,
    tab: str,
    *,
    organization_slug: str | None = None,
) -> bool:
    tab_key = str(tab or "").strip().lower()
    entity = normalize_worker_entity(worker_entity, organization_slug=organization_slug)
    if entity is None:
        entity = ENTITY_NONE
    if tab_key == ENTITY_COMBINED:
        return entity != ENTITY_NONE
    if entity == ENTITY_NONE:
        return False
    if entity == ENTITY_SHARED:
        allowed = set(entities_for_organization(organization_slug, is_privileged=False))
        allowed.discard(ENTITY_COMBINED)
        return tab_key in allowed
    return entity == tab_key


def shift_matches_entity_tab(
    shift_entity: str | None,
    tab: str,
    *,
    organization_slug: str | None = None,
) -> bool:
    tab_key = str(tab or "").strip().lower()
    entity = normalize_shift_entity(shift_entity, organization_slug=organization_slug)
    if tab_key == ENTITY_COMBINED:
        return entity is not None
    if not entity:
        return False
    return entity == tab_key


def worker_allows_shift_entity(
    worker_entity: str | None,
    shift_entity: str | None,
    *,
    organization_slug: str | None = None,
) -> bool:
    worker = normalize_worker_entity(worker_entity, organization_slug=organization_slug) or ENTITY_NONE
    shift = normalize_shift_entity(shift_entity, organization_slug=organization_slug)
    if worker == ENTITY_NONE or not shift:
        return False
    if worker == ENTITY_SHARED:
        return True
    return worker == shift


def assert_same_organization(record_org_id: Any, tenant_org_id: int) -> None:
    if int(record_org_id or 0) != int(tenant_org_id):
        raise ValueError("cross-organization access denied")


def is_privileged_user(user_roles: Sequence[str] | None) -> bool:
    roles = {str(r).upper() for r in (user_roles or [])}
    return bool(roles & PRIVILEGED_ROLES)


def ensure_business_entity_column(cursor) -> None:
    """Add payroll_worker_profiles.business_entity when missing (mirrors business_entity_v1.sql)."""
    from backend.ta_helpers import table_has_column

    if table_has_column(cursor, "payroll_worker_profiles", "business_entity"):
        return
    cursor.execute(
        """
        ALTER TABLE payroll_worker_profiles
        ADD COLUMN business_entity VARCHAR(32) NULL DEFAULT NULL AFTER can_work_both
        """
    )


def entity_scope_payload(
    organization_id: int,
    organization_slug: str | None,
    user_roles: Sequence[str] | None,
) -> dict[str, Any]:
    slug = normalize_org_slug(organization_slug)
    privileged = is_privileged_user(user_roles)
    tabs = entities_for_organization(slug, is_privileged=privileged)
    return {
        "organization_id": int(organization_id),
        "organization_slug": slug,
        "default_entity": default_entity_for_org(slug),
        "entity_tabs": tabs,
        "active_entity_hint": default_entity_for_org(slug),
        "combined_is_admin_view": privileged,
    }
